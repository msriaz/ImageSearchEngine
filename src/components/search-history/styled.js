import styled from 'styled-components/native';

export const StyledTextInput = styled.TextInput`
  border: 1px;
  border-color: ${({theme}) => theme.color.lightGray};
  width: ${({theme}) => theme.windowWidth * 0.75}px;
  padding: ${({theme}) => theme.spacing.small}px;
`;

export const StyledButton = styled.Text``;

export const StyledWrapper = styled.View`
  justify-content: space-between;
  margin: ${({theme}) => theme.spacing.small}px;
`;

export const StyledHeader = styled.Text`
  font-size: 18px;
  font-weight: bold;
  margin-bottom: ${({theme}) => theme.spacing.small}px;
`;

export const HeaderWrapper = styled.View`
  justify-content: space-between;
  margin: ${({theme}) => theme.spacing.small}px;
  flex-direction: row;
  align-items: center;
`;

export const StyledText = styled.Text`
  font-size: 12px;
  font-weight: 200;
  /* margin-left: ${({theme}) => theme.spacing.default}px; */
`;
