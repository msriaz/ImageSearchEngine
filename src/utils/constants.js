import {color as colors} from '../theme';
import {generateColorPalleteConstants} from './index';

export const APP_COLORS = generateColorPalleteConstants(colors);

export const STORAGE_KEYS = {
  HISTORY: 'history',
};
