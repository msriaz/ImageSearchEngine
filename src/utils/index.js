import {useEffect, useRef} from 'react';
import snakeCase from 'lodash/snakeCase';
/**
 *
 * @param {*} value
 * @returns previous state
 */
export const usePrevious = value => {
  const ref = useRef();
  useEffect(() => {
    ref.current = value;
  });
  return ref.current;
};

/**
 *
 * @param {*} imgData
 * @returns image url to render
 */
export const getImageUrl = imgData => {
  const {farm, server, id, secret} = imgData;
  return `https://farm${farm}.static.flickr.com/${server}/${id}_${secret}.jpg`;
};

/**
 * generate color consts from color pallette
 * @param {*} colors
 */
export const generateColorPalleteConstants = colors => {
  const colorsList = {};
  Object.keys(colors).forEach(colorKey => {
    colorsList[snakeCase(colorKey).toUpperCase()] = colors[colorKey];
  });
  return colorsList;
};

/**
 *
 * @param {*} funcName
 * @param {*} error
 */
export const reportError = (funcName, error) => {
  console.log(`Error in ${funcName}`, error);
};
