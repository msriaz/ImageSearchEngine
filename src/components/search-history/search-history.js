import React, {useEffect, useState} from 'react';
import {
  StyledHeader,
  StyledWrapper,
  StyledButton,
  HeaderWrapper,
  StyledText,
} from './styled';
import {load, clear} from '../../utils/storage';
import {STORAGE_KEYS} from '../../utils/constants';
import {reportError} from '../../utils';
import {HistoryList} from './history-iist';

export const SearchHistory = ({onPressHistoryItem, searchQuery}) => {
  const [history, setHistory] = useState([]);

  useEffect(() => {
    loadHistory();
  }, [searchQuery]);

  const loadHistory = async () => {
    try {
      const savedHistory = await load(STORAGE_KEYS.HISTORY);
      console.log('load more history');
      if (savedHistory) {
        setHistory(savedHistory);
      }
    } catch (error) {
      reportError('loadHistory', error);
    }
  };

  const onClearHistory = async () => {
    try {
      await clear();
      setHistory([]);
    } catch (error) {
      reportError('onClearHistory', error);
    }
  };

  return history?.length ? (
    <StyledWrapper>
      <HeaderWrapper>
        <StyledHeader>Search History</StyledHeader>
        <StyledButton onPress={onClearHistory}>Clear History</StyledButton>
      </HeaderWrapper>
      <HistoryList data={history} onPressHistoryItem={onPressHistoryItem} />
    </StyledWrapper>
  ) : (
    <StyledWrapper>
      <StyledText> No previous history found </StyledText>
    </StyledWrapper>
  );
};
