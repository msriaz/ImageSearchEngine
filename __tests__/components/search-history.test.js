import React from 'react';
import {SearchHistory} from '../../src/components/search-history';
import {render} from '../utils/theme-wrapper';

const onPressMock = jest.fn();

describe('Search History Component', () => {
  it('should match the Search History  snapshot', () => {
    const compTree = render(<SearchHistory onPressHistoryItem={onPressMock} />);
    expect(compTree.toJSON()).toMatchSnapshot();
  });
});
