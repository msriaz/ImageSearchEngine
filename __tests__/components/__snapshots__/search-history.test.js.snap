// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search History Component should match the Search History  snapshot 1`] = `
<View
  style={
    Array [
      Object {
        "justifyContent": "space-between",
        "marginBottom": 8,
        "marginLeft": 8,
        "marginRight": 8,
        "marginTop": 8,
      },
    ]
  }
>
  <Text
    style={
      Array [
        Object {
          "fontSize": 18,
          "fontWeight": "bold",
        },
      ]
    }
  >
    History
  </Text>
  <RCTScrollView
    data={Array []}
    getItem={[Function]}
    getItemCount={[Function]}
    horizontal={true}
    keyExtractor={[Function]}
    onContentSizeChange={[Function]}
    onLayout={[Function]}
    onMomentumScrollBegin={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={50}
    stickyHeaderIndices={Array []}
    viewabilityConfigCallbackPairs={Array []}
  >
    <View />
  </RCTScrollView>
</View>
`;
